File: generated_code/GeminiPro/Java/048_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree.
 * This class implements a set of integers.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation.
     * Brings the node with the given key (or the last accessed node if the key is not found)
     * to the root of the tree rooted at 'node'.
     * This is a recursive, bottom-up splay implementation.
     *
     * @param node The root of the tree/subtree to splay.
     * @param key The key to search for and bring to the root.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left Left)
            if (key < node.left.key) {
                // First recursively bring the key as root of left-left
                node.left.left = splay(node.left.left, key);
                // Do first rotation for node, second rotation is done after else
                node = rightRotate(node);
            }
            // Zig-Zag (Left Right)
            else if (key > node.left.key) {
                // First recursively bring the key as root of left-right
                node.left.right = splay(node.left.right, key);
                // Do first rotation for node.left
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Do second rotation for node
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key lies in the right subtree
        else {
            // Key is not in the tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right Left)
            if (key < node.right.key) {
                // Bring the key as root of right-left
                node.right.left = splay(node.right.left, key);
                // Do first rotation for node.right
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > node.right.key) {
                // Bring the key as root of right-right and do first rotation
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }

            // Do second rotation for node
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Searches for a key in the tree.
     * If the key is found, it is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree.
     * If the key already exists, the tree is splayed on that key but not modified.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root
        root = splay(root, key);

        // If key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // The new node becomes the new root. The old root and its subtrees
        // are attached as children of the new root.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     * If the key is not found, the tree is splayed on the last accessed node.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (if it exists) to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now the node to be deleted is the root.
        // We need to join its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // This maximum element has no right child.
            leftSubtree = splay(leftSubtree, key); // or splay for a very large key

            // Attach the original right subtree as the right child of the new root.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/048_SplayTree.py
class SplayTree:
    """
    An implementation of a Splay Tree that acts as a dictionary-like set for integers.
    Splay trees are self-balancing binary search trees with the additional property
    that recently accessed elements are quick to access again.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on the subtree rooted at node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the subtree rooted at node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (x is left child, parent is right child)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.
        Performs the splaying operation on the accessed node or its would-be parent.
        Returns True if the key is found, False otherwise.
        """
        node = self.root
        last_visited = None
        found = False

        while node:
            last_visited = node
            if key == node.key:
                found = True
                break
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        if node:  # Key was found
            self._splay(node)
        elif last_visited:  # Key not found, splay the last visited node
            self._splay(last_visited)

        return found

    def insert(self, key):
        """
        Inserts a key into the tree.
        If the key already exists, the tree remains unchanged but the existing
        node with the key is splayed to the root.
        """
        # If tree is empty, create a new root
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the closest node to the root
        self.search(key)

        # If key is already at the root, we're done
        if self.root.key == key:
            return

        # Otherwise, split the tree and insert the new node as the root
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.parent = new_node
            self.root.left = None
        else: # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.parent = new_node
            self.root.right = None
        
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.
        If the key is not found, the tree structure may change due to splaying,
        but no node is removed.
        """
        # Splay the node to be deleted (or its parent) to the root
        self.search(key)

        # If the key is not at the root, it wasn't in the tree
        if not self.root or self.root.key != key:
            return

        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If no left subtree, the right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Detach the left subtree
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this max node to the root of the left subtree
            # Temporarily set root to perform splay within the subtree
            self.root = left_subtree
            self._splay(max_node)

            # The new root (max_node) now has the right_subtree as its right child
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root